### ISSUE TITLE
Floodlight wire connection weirdness

### ISSUE THREAD

**Issue by theunkn0wn1** [BluSunrize/ImmersiveEngineering#551]
Issue: It is possible to place LV wires through walls with a certain placement of floodlights. Pictures below show the setup:

![image](https://cloud.githubusercontent.com/assets/3110986/10867364/56955dc0-801b-11e5-8c85-968486e54c41.png)

![image](https://cloud.githubusercontent.com/assets/3110986/10867369/7b4e5b76-801b-11e5-8727-aef1eb206c97.png)
![image](https://cloud.githubusercontent.com/assets/3110986/10867372/8eb6fbe6-801b-11e5-9565-81efbfa844a2.png)

And a practical implementation...
![image](https://cloud.githubusercontent.com/assets/3110986/10867385/3502ae8c-801c-11e5-9a7a-44743c2d5243.png)

By starting at the lower flood light and connecting it to any of the upper ones, a valid connection is formed. This is unexpected, as there is no LoS between the connections due to the existence of a 2 meter thick solid wall between them.

IE build cf77da7
Bare-minimum MC instance.

**Comment by cobra** [comment]
That's interesting. I'll investigate, I was the last one to touch the raytracing code afaik - but floodlights didn't exist then.

**Comment by cobra** [comment]
I can't reproduce all of these cases, but your last one works in some orientations.
![](https://cloud.githubusercontent.com/assets/163331/10868214/0791a1ea-8087-11e5-850c-5b7a069e0981.png)
North is up, the lanterns with the red clay are the ones I can connect from the floodlight in the middle.

**Comment by cobra** [comment]
The raytraceOffsets are a bit off, but even after finding these points I'm still confused how minecraft's algorithm doesn't collide with all the stuff which should definitely block. The raytrace is always returning null in these cases.
I once again wish I could just render the raytrace path. If anyone wants to join the bug hunt, feel free to do so.

**Comment by BluSunrize** [comment]
I'll file this at low priority for now.

In fact, I'd rather get your PRs in @cobra. I was kinda hoping to maybe wrap up IE today or in the next few days.
Also, @cobra, you should really hop onto IRC some time (#BluSunrize on esper.net). It's just so much more convenient for discussions. =P

**Comment by cobra** [comment]
I'm 99% sure that this bug can be found somewhere in the getRaytraceOffset methods for both the lantern and the floodlight. Which are looking a bit scary. (hey, Halloween is over!)

**Comment by BluSunrize** [comment]
Yeah they look like that because they have 4 different "connectors" ^^

**Comment by cobra** [comment]
Out of curiosity I tested the old code again: the vanilla raytracer is checking a single block without hitting anything for the floodlight→floodlight case, which is definitely a bug. But as we are now using the new raytracer by @malte0811, this bug doesn't affect IE any more.

**Comment by cobra** [comment]
Update: it was the 1% chance - the vanilla raytracer is broken and doesn't check everything in exactly this specific way. It has now been replaced by a custom implementation by malte0811, which has fixed this particular bug. The raytraceOffsets still need a bit of work, though. I already removed the y change for the electric lantern and made the floodlight respect it's base orientation, but that wasn't fully finished. Will do that today, there's not much left to check.
